File: cpp/4_largest_palindrome_product.cpp
#include <iostream>

bool checkPalindrome(int n) {
    int original = n;
    int reversed = 0;

    while (n != 0) {
        reversed = reversed * 10 + n % 10;
        n /= 10;
    }

    return original == reversed;
}

int main() {
    int largest = 0;

    for (int i = 100; i < 1000; i++) {
        for (int j = i; j < 1000; j++) {
            int result = i * j;
            if (checkPalindrome(result)) {
                if (result > largest) {
                    largest = result;
                }
            }
        }
    }

    std::cout << largest << std::endl;
}

File: lua/4_largest_palindrome_product.lua
function checkPalindrome(n)
    local original, reversed = n, 0

    while n ~= 0 do
        reversed = reversed * 10 + n % 10
        n = math.floor(n / 10)
    end

    return original == reversed
end

local largest = 0;

for i = 100, 999 do
    for j = i, 999 do
        local result = i * j
        if checkPalindrome(result) then
            if result > largest then
                largest = result
            end
        end
    end
end

print(largest)